compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -642,12 +642,18 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
     }
     val leaks = escapingRefs(tree, localSyms)
     if (leaks.isEmpty) tree
-    else if (isFullyDefined(pt, ForceDegree.none)) ascribeType(tree, pt)
     else if (!forcedDefined) {
       fullyDefinedType(tree.tpe, "block", tree.pos)
-      val tree1 = ascribeType(tree, avoid(tree.tpe, localSyms))
-      ensureNoLocalRefs(tree1, pt, localSyms, forcedDefined = true)
-    } else
+      val avoidingType = avoid(tree.tpe, localSyms)
```

Comment:
The comment on top of this method needs to be updated now that the order of operations changed.